File: Python/7 - Funciones.py
# Definir funciones con def

# 1. Función básica sin parámetros
def saludar():
    """Imprime un mensaje de bienvenida."""
    print("¡Hola, bienvenido la clase de funciones de Python!")

# 2. Función con parámetros
def funcion (x):
    return x + 5

# 3. Funcion con varios parametros
def funcion2 (x, y):
    z = x - y   # Las variables que se crean en una funcion solo exsiten en la funcion
    return z

# 4. Función con múltiples valores de retorno
def calcular_estadisticas(numeros):
    """Calcula la suma, promedio, máximo y mínimo de una lista de números."""
    suma = sum(numeros)
    promedio = suma / len(numeros)
    maximo = max(numeros)
    minimo = min(numeros)
    return suma, promedio, maximo, minimo

# 5. Funcion que devuelve el caracter mas repetido de un String
def caracter_mas_repetido(cadena):
    if not cadena:
        return None
    
    # Crear un diccionario para contar la frecuencia de cada carácter
    frecuencias = {}
    for char in cadena:
        if char in frecuencias and char != ' ':
            frecuencias[char] += 1
        else:
            frecuencias[char] = 1
    
    # Encontrar el carácter con la mayor frecuencia
    max_char = max(frecuencias, key=frecuencias.get)
    
    return max_char

# 6. Función con parámetros por defecto
def saludar_persona(nombre="Invitado"):
    """Saluda a una persona usando un valor por defecto si no se proporciona nombre."""
    return f"Hola {nombre}!"

# 7. Función con argumentos variables (*args)
def sumar_numeros(*numeros):
    """Suma una cantidad variable de números."""
    return sum(numeros)

# 8. Función con argumentos de palabra clave variables (**kwargs)
def info_persona(**datos):
    """Muestra información de una persona con pares clave-valor variables."""
    for clave, valor in datos.items():
        print(f"{clave}: {valor}")

# 9. Función con decorador para medir tiempo de ejecución
from time import time
def medir_tiempo(funcion):
    def wrapper(*args, **kwargs):
        inicio = time()
        resultado = funcion(*args, **kwargs)
        fin = time()
        print(f"Tiempo de ejecución: {fin - inicio} segundos")
        return resultado
    return wrapper

# 10. Decorador aplicado a una función
@medir_tiempo # Decorador para medir tiempo de ejecución. la función se ejecuta antes de la decoradora
def operacion_lenta():
    return sum(range(10000000))

# Programa principal
if __name__ == '__main__':
    saludar() # Invocar la funcion
    print(f'Comienzo del programa {funcion(5)}')
    
    valor1 = 10
    valor2 = 4
    valor3 = funcion2(valor1, valor2)
    print(valor3)

    vector = [15, 23, 20, 11, 9, 16, 15, 22, 30, 17]
    suma, promedio, max_num, min_num = calcular_estadisticas(vector)
    print(f"Suma: {suma}, Promedio: {promedio:.2f}, Máximo: {max_num}, Mínimo: {min_num}")
    print(f'Estadisticas: {calcular_estadisticas(vector)}')

    x = "Mi nombre es Máximo decimo Meridio, comandante de los ejércitos del norte, general de las legiones Félix. leal, sirviente del único emperador Marco Aurelio, padre de un hijo asesinado, esposo de una esposa asesinada. Y juro que me vengaré. en esta vida o en la otra."
    resultado = caracter_mas_repetido(x)
    print(f"El carácter que más se repite en '{x}' \nes '{resultado}'")

    suma = sumar_numeros(1, 2, 3)  # Suma números fijos
    print(f"La suma de (1, 2, 3) es: {suma}")  # Suma números variables
    
    info_persona(nombre="Juan", edad=30, ciudad="Madrid")  # Kwargs ejemplo

    # Los decoradores sirven para añadir funcionalidades a funciones ya definidas
    # sin modificar su código. Se utilizan para añadir funcionalidades a funciones ya definidas
    operacion_lenta()

File: Ruby/7 - Funciones.rb
# Definir funciones con def

# 1. Función básica sin parámetros
def saludar
  puts "¡Hola, bienvenido al curso de Ruby!"
end

# 2. Función con parámetros
def funcion(x)
  x + 5  # En Ruby, la última expresión es implícitamente retornada
end

# 3. Función con varios parámetros
def funcion2(x, y)
  z = x - y   # Las variables que se crean en una función solo existen en la función
  z  # retorno implícito
end

# 4. Función con múltiples valores de retorno
def calcular_estadisticas(numeros)
  return [0, 0.0, nil, nil] if numeros.empty?  # Manejo de array vacío

  suma = numeros.sum
  promedio = suma.to_f / numeros.length
  maximo = numeros.max
  minimo = numeros.min
  [suma, promedio, maximo, minimo]  # Ruby retorna un array
end

# 5. Función que devuelve el caracter más repetido de un String
def caracter_mas_repetido(cadena)
  return nil if cadena.nil? || cadena.empty?

  # Crear un hash para contar la frecuencia de cada carácter
  frecuencias = Hash.new(0)
  cadena.each_char do |char|
    frecuencias[char] += 1 unless char == ' '  # Ignoramos espacios
  end

  # Encontrar el carácter con la mayor frecuencia
  frecuencias.max_by { |_, count| count }&.first
end

# 6. Función con parámetros por defecto
def saludar_persona(nombre = "Invitado")
  "Hola #{nombre}!"
end

# 7. Función con argumentos variables (*args)
def sumar_numeros(*numeros)
  numeros.sum
end

# 8. Función con argumentos de palabra clave variables (**kwargs)
def info_persona(**datos)
  datos.each do |clave, valor|
    puts "#{clave}: #{valor}"
  end
end

# 9. Módulo para medir tiempo de ejecución
module MedirTiempo
  def medir_tiempo(metodo)
    alias_method "original_#{metodo}", metodo

    define_method(metodo) do |*args, &block|
      inicio = Time.now
      resultado = send("original_#{metodo}", *args, &block)
      fin = Time.now
      puts "Tiempo de ejecución de #{metodo}: #{fin - inicio} segundos"
      resultado
    end
  end
end

# 10. Clase con método que usa el módulo de medición de tiempo
class Operaciones
  extend MedirTiempo  # Usamos extend en lugar de include para métodos de clase

  def operacion_lenta
    (0..10_000_000).sum
  end

  medir_tiempo :operacion_lenta
end

# Programa principal
if __FILE__ == $PROGRAM_NAME
  puts "\n--- Función básica ---"
  saludar

  puts "\n--- Función con parámetros ---"
  puts "Resultado de funcion(5): #{funcion(5)}"

  puts "\n--- Función con varios parámetros ---"
  valor1 = 10
  valor2 = 4
  valor3 = funcion2(valor1, valor2)
  puts "Resultado de funcion2(#{valor1}, #{valor2}): #{valor3}"

  puts "\n--- Función con múltiples retornos ---"
  vector = [15, 23, 20, 11, 9, 16, 15, 22, 30, 17]
  suma, promedio, max_num, min_num = calcular_estadisticas(vector)
  puts "Suma: #{suma}, Promedio: #{'%.2f' % promedio}, Máximo: #{max_num}, Mínimo: #{min_num}"
  puts "Estadísticas completas: #{calcular_estadisticas(vector)}"

  puts "\n--- Función de caracteres ---"
  texto = "Mi nombre es Máximo decimo Meridio, comandante de los ejércitos del norte"
  resultado = caracter_mas_repetido(texto)
  puts "El carácter que más se repite en el texto es '#{resultado}'"

  puts "\n--- Función con parámetro por defecto ---"
  puts saludar_persona
  puts saludar_persona("Carlos")

  puts "\n--- Función con argumentos variables ---"
  puts "Suma de (1, 2, 3): #{sumar_numeros(1, 2, 3)}"
  puts "Suma de (4, 5, 6, 7): #{sumar_numeros(4, 5, 6, 7)}"

  puts "\n--- Función con kwargs ---"
  info_persona(nombre: "Juan", edad: 30, ciudad: "Madrid")

  puts "\n--- Función con medición de tiempo ---"
  operaciones = Operaciones.new
  puts "Resultado de operación lenta: #{operaciones.operacion_lenta}"
end